app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt

```diff
@@ -44,32 +54,46 @@ class TokenAdapter(
 
     override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
         when (holder) {
-            is TokenViewHolder -> holder.onBind(data[position] as HomeElementItem.Shown, isZerosHidden)
-            is TokenHiddenViewHolder -> holder.onBind(data[position] as HomeElementItem.Hidden, isZerosHidden)
-            is TokenButtonViewHolder -> holder.onBind(data[position] as HomeElementItem.Action)
-            is BannersViewHolder -> holder.onBind(data[position] as HomeElementItem.Banners)
-            is HeaderViewHolder -> holder.onBind(data[position] as HomeElementItem.Title)
+            is TokenViewHolder -> holder.onBind(data[position] as Shown, isZerosHidden)
+            is TokenHiddenViewHolder -> holder.onBind(data[position] as Hidden, isZerosHidden)
+            is TokenButtonViewHolder -> holder.onBind(data[position] as Action)
+            is BannersViewHolder -> holder.onBind(data[position] as Banners)
+            is HeaderViewHolder -> holder.onBind(data[position] as Title)
         }
     }
 
-    private fun mapGroups(tokens: List<HomeElementItem>, state: VisibilityState): List<HomeElementItem> =
-        tokens
-            .groupBy { it is HomeElementItem.Hidden }
-            .map { (isHidden, list) -> mapTokensGroup(isHidden, list, state) }
-            .flatten()
-
-    private fun mapTokensGroup(
-        isHidden: Boolean,
-        list: List<HomeElementItem>,
-        state: VisibilityState
-    ) = list.toMutableList().apply {
-        if (isHidden) {
-            add(0, HomeElementItem.Action(state))
-        } else {
-            add(0, HomeElementItem.Title(R.string.home_tokens))
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int, payloads: MutableList<Any>) {
+        val fields = (payloads.firstOrNull() as? Set<*>)?.filterIsInstance<String>()
```

Comment:
god damn
that's why we need delegates  😃

Reply:
I made a minor change. 
It should look better